ddtrace/contrib/internal/django/templates.py

```diff
@@ -0,0 +1,129 @@
+from types import FunctionType
+from types import ModuleType
+from types import TracebackType
+import typing
+from typing import Optional
+from typing import Type
+from typing import TypeVar
+
+import ddtrace
+from ddtrace import config
+from ddtrace._trace.pin import Pin
+from ddtrace.ext import http
+from ddtrace.internal import core
+from ddtrace.internal.compat import maybe_stringify
+from ddtrace.internal.constants import COMPONENT
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils.importlib import func_name
+from ddtrace.internal.wrapping.context import WrappingContext
+
+
+T = TypeVar("T")
+
+log = get_logger(__name__)
+
+
+class DjangoTemplateWrappingContext(WrappingContext):
+    """
+    A context for wrapping django.template.base:Template.render method.
+    """
+
+    def __init__(self, f: FunctionType, django: ModuleType) -> None:
+        super().__init__(f)
+        self._django = django
+
+    @classmethod
+    def instrument_module(cls, django_template_base: ModuleType) -> None:
+        """
+        Instrument the django template base module to wrap the render method.
+        """
+        if not config.django.instrument_templates:
+            return
+
+        # Wrap the render method of the Template class
+        Template = getattr(django_template_base, "Template", None)
+        if not Template or not hasattr(Template, "render"):
+            return
+
+        import django
+
+        cls(typing.cast(FunctionType, Template.render), django).wrap()
+
+    @classmethod
+    def uninstrument_module(cls, django_template_base: ModuleType) -> None:
+        # Unwrap the render method of the Template class
+        Template = getattr(django_template_base, "Template", None)
+        if not Template or not hasattr(Template, "render"):
+            return
+
+        if cls.is_wrapped(Template.render):
+            ctx = cls.extract(Template.render)
+            ctx.unwrap()
+
+    def __enter__(self) -> "DjangoTemplateWrappingContext":
+        super().__enter__()
+
+        if not config.django.instrument_templates:
+            return self
+
+        # Get the template instance (self parameter of the render method)
+        # Note: instance is a django.template.base.Template
+        instance = self.get_local("self")
+
+        # Extract template name
+        template_name = maybe_stringify(getattr(instance, "name", None))
+        if template_name:
+            resource = template_name
+        else:
+            resource = "{0}.{1}".format(func_name(instance), self.__wrapped__.__name__)
+
+        # Build tags
+        tags = {COMPONENT: config.django.integration_name}
+        if template_name:
+            tags["django.template.name"] = template_name
+
+        engine = getattr(instance, "engine", None)
+        if engine:
+            tags["django.template.engine.class"] = func_name(engine)
+
+        # Create the span context
+        pin = Pin.get_from(self._django)
```

Comment:
can't the pin be stored on the wrapping context instead?

Reply:
1. I want to get rid of Pin entirely for everything everywhere
2. Doing that breaks tests.... I decided to take a "get this working without touching any tests" approach for v1